implement rocksdb tcp server and update client

ertugrulaypek opened the pull request:
**Changes**

- Rocksdb server is implemented which gets requests from multiple clients concurrently over tcp. Boost async io library is used.
- Rocksdb client is updated in controllers. Boost tcp client is used.

**Notes**

- There is a warning due to recursion in `do_read` method [here](https://github.com/dimstav23/GDPRuler/compare/separate-rocksdb-client-server?expand=1#diff-52aa306fcfb9d7b567403b0bc3631bccb1008c5a7b20fb14663144d8ec2b4bbaR37). I think it is unavoidable if we use boost lib since the callback we bind should be called multiple times before closing a connection. Otherwise, only a part of the stream is received. Please let me know if there is a way to avoid this.

**Tests**

- Tested with workloads a-d and f. All of them are working with both `native` and `gdpr` controllers.
- Tested multiple workloads simultaneously - no problems.

dimstav23 commented on controller/source/rocksdb_server/message.hpp:
```diff
@@ -0,0 +1,152 @@
+
+private:
+  bool m_is_success {false};
+  std::string m_data;
```
Maybe add a comment indicating the current message format for get/put/delete with an example?
That would be helpful for future adaptations. @ertugrulaypek

ertugrulaypek replied:
added docstring with examples

dimstav23 commented on controller/source/rocksdb_server/server.cpp:
```diff
@@ -0,0 +1,95 @@
+
+using boost::asio::ip::tcp;
+
+class session : public std::enable_shared_from_this<session> {
```
Add a comment about what the session object refers to (because it's a generic term)
@ertugrulaypek

dimstav23 replied:
We might need a similar concept for redis client so in a next version we might need to generlize the `session` term.

ertugrulaypek replied:
added docstrings to `session`, `rocksdb_proxy` and `rocksdb_server` classes.

dimstav23 commented on controller/source/kv_client/rocksdb.hpp:
```diff
@@ -1,65 +1,93 @@
+    assert(host_port_splits.size() == 2 && "DB server address must be in <host>:<port> format!");
+
+    // Resolve the host and port to an endpoint
+    using boost::asio::ip::tcp;
```
maybe move this using directive on the preamble of the file? would make it cleaner. @ertugrulaypek

ertugrulaypek replied:
I put it here because this is the only scope that needs the namespace. I did not want to pollute our namespace. But if that does not make sense to you, I can move it

dimstav23 replied:
It totally makes sense. Just looked weird compared to other files we have. You can leave it as it is.

dimstav23 commented on controller/source/rocksdb_server/server.cpp:
```diff
@@ -0,0 +1,95 @@
+        response_message response = m_rocksdb_proxy->execute(query);
+        std::string raw_response = response.serialize();
+        boost::asio::write(m_socket, boost::asio::buffer(raw_response));
+        do_read();
```
Couldn't this recursion be done with a while loop? (it's fine.. just to avoid the warning)

dimstav23 commented on controller/source/rocksdb_server/server.cpp:
```diff
@@ -0,0 +1,95 @@
+        std::string raw_query(boost::asio::buffers_begin(m_buffer.data()), boost::asio::buffers_end(m_buffer.data()) - 1);
+        m_buffer.consume(length);
+        query_message query = query_message::deserialize(raw_query);
+        response_message response = m_rocksdb_proxy->execute(query);
```
Since the function also `executes` the query, maybe `do_read()` is not the best naming.

ertugrulaypek replied:
yes, I agree. Changed it with `handle_read`

dimstav23 commented on controller/source/rocksdb_server/server.cpp:
```diff
@@ -0,0 +1,95 @@
+
+class rocksdb_server {
+public:
+  rocksdb_server(boost::asio::io_service& io_service,
```
Could  be helpful to add a comment about the io_service, and generally the choice of boost lib types to avoid looking the documentation (minor)

ertugrulaypek replied:
done

dimstav23 commented on controller/source/rocksdb_server/server.cpp:
```diff
@@ -0,0 +1,95 @@
+    assert(argc == 3 && "Usage: ./rocksdb_server <port> <db_path>");
+    boost::asio::io_service io_service;
+    rocksdb_server rocksdb_server(io_service, static_cast<uint16_t>(std::stoul(args[1])), args[2]);
+    io_service.run();
```
same here.. a comment would be helpful

ertugrulaypek replied:
done

dimstav23 reviewed (commented):
An overall comment about the server side:
Currently, the server accepts connections and creates sessions that execute these recursions..
Wouldn't it be meaningful to also add a thread assignment from a pool of threads (let's say 16 threads or ,even better, taken as a parameter) or is it already acheved and I missed it?

dimstav23 commented on controller/source/rocksdb_server/message.hpp:
```diff
@@ -99,6 +112,22 @@ class query_message
+ * 
+ * The raw message contains two fields: is_success, data.
+ * is_success represents the result of an operation.
+ * data represents the value retrieved in case of a successful put operation, 
```
maybe a successful *get* operation?

ertugrulaypek replied:
yeap, nice catch. fixed it :+1:

ertugrulaypek commented:
> An overall comment about the server side: Currently, the server accepts connections and creates sessions that execute these recursions.. Wouldn't it be meaningful to also add a thread assignment from a pool of threads (let's say 16 threads or ,even better, taken as a parameter) or is it already acheved and I missed it?

Thanks for your review! I added a new command line argument to the program and created threads to run async handlers as suggested [here](https://theboostcpplibraries.com/boost.asio-scalability).

Additionally, I fixed all the comments and added a README.md file to the directory. Let me know if there are any other concerns
